Add mozfun.norm.fenix_app_info UDF

jklukas opened the pull request:
Relevant to #947

We will use this function for how we present Firefox Preview and Fenix in GUD.

acmiyaguchi commented on udf/extract_canonical_glean_app_info.sql:
```diff
@@ -0,0 +1,71 @@
+/*
```
Should this UDF live inside of mozfun instead?

jklukas replied:
I would indeed like to get it into `mozfun` eventually, but I see a few blockers:

- It depends on `udf.fenix_build_to_datetime`, which we'd also need to move to mozfun
- It currently only supports Fenix
- This UDF still feels a bit volatile; it's not clear that the two arguments we have in here now are going to be sufficient long-term to disambiguate app names and channels

So I'm thinking of this a bit as an incubation stage before committing to an interface and a long-term name in mozfun. And I'm definitely open to suggestions on naming.

acmiyaguchi replied:
It seems like it would be better to just commit with a single location for this UDF, especially if it's intended to be used across projects. FWIW, it would be useful to have `fenix_build_to_datetime` in mozfun too.

acmiyaguchi commented on udf/extract_canonical_glean_app_info.sql:
```diff
@@ -0,0 +1,71 @@
+      STRUCT('Fenix', 'nightly', 'org.mozilla.fenix')
+    )
+  ELSE
+    ERROR("Bad data")
```
Is there a way to test for this case?

jklukas replied:
I meant to fill this out with a better message. But no, I'm not aware of a way at this point to write a test case that exercises this. If we raise an error, the test fails.

acmiyaguchi commented on udf/extract_canonical_glean_app_info.sql:
```diff
@@ -0,0 +1,71 @@
+/*
+
```
Perhaps link to the historical naming doc for more context on naming.

jklukas replied:
Are you referring to https://docs.google.com/document/d/1O_hDT8po5iEQcle62fN_eno4liVhZOhFXnKRBAoBaWk/edit?ts=5f31bb6d#heading=h.z5xwoh2241fq ?

acmiyaguchi replied:
No, I was thinking this spreadsheet: https://docs.google.com/spreadsheets/d/18PzkzZxdpFl23__-CIO735NumYDqu7jHpqllo0sBbPA/edit#gid=0

jklukas commented:
The more time I spend thinking about this, the more uncomfortable I am with releasing this UDF as an interface we recommend folks use. There's just too much uncertainty about what inputs we'll need for future Glean apps (or even for future evolution of Fenix).

I think perhaps a better solution is to push logic into views. In `scripts/generate_views`, we would inject logic such that the user-facing views for Fenix apps would replace `normalized_channel` and would add `normalized_app_name` and `normalized_app_id` fields. We would also propagate those fields to the baseline_clients_daily and baseline_clients_last_seen tables.

There are still some details to work out, but it gives us the flexibility to change what fields we use for calculating app names and channels without impacting users. It also means users don't ever need to know about a UDF; the values are built into the user-facing views.

acmiyaguchi commented:
There are a couple of problems to note with using a view over the stable tables:

* `org_mozilla_fenix` is split across a build_id boundary which is defined within the nested `client_info` field. A view would need to filter on this field requiring a full scan of each partition and would not be able to take advantage of clustering unless explicitly added to the pipeline family metadata.
* The schemas across the products are not guaranteed to the same, so it may not be as simple as unioning the tables. This works the Glam ETL because there the table is flattened in such a way that there a row for each (client, metric name, values).

jklukas commented:
> There are a couple of problems to note with using a view over the stable tables

Your points are valid if we try to make a single view that unifies all fenix apps. I was imagining that we still have the 5 separate app IDs, and we'd be updating the logic for the user-facing views under those 5 datasets. It's still asking a lot for users to have to union together these views.

The `baseline_clients_daily` and `baseline_clients_last_seen` tables are similar to the Glam case where we are generating the tables consistently, and can expect the schemas to match exactly for all glean app ids, thus a union can work. But it would indeed be nice to be able to cluster efficiently on channel.

jklukas commented on mozfun/norm/fenix_app_info/udf.sql:
```diff
@@ -0,0 +1,50 @@
+CREATE OR REPLACE FUNCTION norm.fenix_app_info(app_id STRING, app_build_id STRING)
+RETURNS STRUCT<app_name STRING, channel STRING, app_id STRING> AS (
```
Changed the `canonical_app_id` field to just `app_id` for simplicity. Particularly because the function name no longer uses "canonical".

jklukas commented on mozfun/norm/fenix_app_info/udf.sql:
```diff
@@ -0,0 +1,50 @@
+    IF(
+      -- See udf.fenix_build_to_datetime for info on build_id format;
+      -- the build_id constant here corresponds to 2020-07-03
+      app_build_id < '21850000',
```
I took out the call to `udf.fenix_build_to_datetime` both for performance and to avoid needing a refactor of `udf.fenix_build_to_datetime` within the scope of this PR.

jklukas commented:
I've moved the function to `mozfun` and changed the name to be Fenix-specific.

Basically, I'm decreasing the overall scope here so that we can meet the immediate need of allowing data users to consistently determine channel of Fenix apps. We can defer the larger problem of how to configure relationships between glean app_ids consistently in ETL.
